### PR TITLE
eth: broadcast block to static and trusted peers as well

### PR DESCRIPTION
**Pull request by pratikspatil024**
# Description

The block is broadcasted to only Sqrt(peers) in geth. We observed that in case of validator-sentry architecture, the validator, sometime, might not receive a block which will force the validator to mine on his own, resulting in a reorg.

In this PR, we are making sure that the block will be broadcasted to all the static and trusted peers of a particular node.

# Changes

- [x] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)
- [ ] Changes only for a subset of nodes

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Nodes audience

In case this PR includes changes that must be applied only to a subset of nodes, please specify how you handled it (e.g. by adding a flag with a default value...)

# Checklist

- [x] I have added at least 2 reviewer or the whole pos-v1 team
- [x] I have added sufficient documentation in code
- [x] I will be resolving comments - if any - by pushing each fix in a separate commit and linking the commit hash in the comment reply
- [ ] Created a task in Jira and informed the team for implementation in Erigon client (if applicable)
- [ ] Includes RPC methods changes, and the Notion documentation has been updated

# Cross repository changes

- [ ] This PR requires changes to heimdall
    - In case link the PR here:
- [ ] This PR requires changes to matic-cli
    - In case link the PR here:

## Testing

- [ ] I have added unit tests
- [ ] I have added tests to CI
- [ ] I have tested this code manually on local environment
- [ ] I have tested this code manually on remote devnet using express-cli
- [ ] I have tested this code manually on mumbai/amoy
- [ ] I have created new e2e tests into express-cli

### Manual tests

Please complete this section with the steps you performed if you ran manual tests for this functionality, otherwise delete it

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

### CODE REVIEW

File: eth/handler.go [hunk 1]
```diff
@@ -606,13 +606,30 @@ func (h *handler) BroadcastBlock(block *types.Block, propagate bool) {
 			log.Error("Propagating dangling block", "number", block.Number(), "hash", hash)
 			return
 		}
+
+		// These are the static and trusted peers which are not
+		// in `transfer := peers[:int(math.Sqrt(float64(len(peers))))]`
+		staticAndTrustedPeers := []*ethPeer{}
+
+		for _, peer := range peers[int(math.Sqrt(float64(len(peers)))):] {
```
Comment (marcello33) [inline eth/handler.go:614]: Not sure about this, just asking: shouldn't this cycle go over the whole range of peers?

File: eth/handler.go [hunk 2]
```diff
@@ -606,13 +606,30 @@ func (h *handler) BroadcastBlock(block *types.Block, propagate bool) {
...
 		}
 
-		log.Trace("Propagated block", "hash", hash, "recipients", len(transfer), "duration", common.PrettyDuration(time.Since(block.ReceivedAt)))
+		// Send the block to the trusted and static peers
+		for _, peer := range staticAndTrustedPeers {
+			log.Trace("Propagating block to static and trusted peer", "hash", hash, "peer", peer.ID())
+			peer.AsyncSendNewBlock(block, td)
+		}
+
+		log.Trace("Propagated block", "hash", hash, "recipients", len(transfer)+len(staticAndTrustedPeers), "duration", common.PrettyDuration(time.Since(block.ReceivedAt)))
```
Comment (marcello33) [inline eth/handler.go:632]: WDYT about logging the number of `transfer` recipient AND the number of `staticAndTrustedPeers` recipients?

File: eth/handler.go [hunk 3]
```diff
@@ -606,13 +606,30 @@ func (h *handler) BroadcastBlock(block *types.Block, propagate bool) {
...
 		// Send the block to a subset of our peers
 		transfer := peers[:int(math.Sqrt(float64(len(peers))))]
 		for _, peer := range transfer {
 			peer.AsyncSendNewBlock(block, td)
 		}
 
-		log.Trace("Propagated block", "hash", hash, "recipients", len(transfer), "duration", common.PrettyDuration(time.Since(block.ReceivedAt)))
+		// Send the block to the trusted and static peers
+		for _, peer := range staticAndTrustedPeers {
+			log.Trace("Propagating block to static and trusted peer", "hash", hash, "peer", peer.ID())
```
Comment (marcello33) [inline eth/handler.go:628]: Shall we rename `peer` to `peerID`, for clarity?

File: eth/protocols/eth/peer.go [hunk 4]
```diff
@@ -470,6 +470,16 @@ func (p *Peer) RequestTxs(hashes []common.Hash) error {
 	})
 }
 
+// IsTrusted returns whether the peer is a trusted peer or not.
+func (p *Peer) IsTrusted() bool {
+	return p.Info().Network.Trusted
```
Comment (marcello33) [inline eth/protocols/eth/peer.go:475]: here and below, for both functions, no risks of `nil` pointers exceptions, right?

File: eth/handler.go (hunk 1 above)
Comment (pratikspatil024) [reply eth/handler.go:614]: So this loop is to only get those static and trusted peers which are not eligible for the legacy broadcast. The peers we are skipping here are already included below
```
transfer := peers[:int(math.Sqrt(float64(len(peers))))]
```

I have mentioned this in the comment
```
// These are the static and trusted peers which are not
// in `transfer := peers[:int(math.Sqrt(float64(len(peers))))]`
staticAndTrustedPeers := []*ethPeer{}
```

File: eth/handler.go (hunk 3 above)
Comment (pratikspatil024) [reply eth/handler.go:630]: Actually `peer` here is the `*ethPeer` structure. `peerID` would be `peer.ID()`

File: eth/handler.go (hunk 2 above)
Comment (pratikspatil024) [reply eth/handler.go]: Makes sense. I refactored the log into 2 parts to keep the geth code unchanged. Does it makes sense?

File: eth/protocols/eth/peer.go (hunk 4 above)
Comment (pratikspatil024) [reply eth/protocols/eth/peer.go:475]: Yes, I think we should be fine here. Checked the `Info()` function and other occurances of it being called.

**Review by ssandeep — APPROVED** [review]

File: eth/handler.go (hunk 3 above)
Comment (marcello33) [reply eth/handler.go:630]: Yeah, I meant in the trace log, where you print `peer` as `peer.ID()`
Anyway, minor one

**Review by marcello33 — APPROVED** [review]

**Review by cffls — APPROVED** [review]
LGTM

**Review by manav2401 — APPROVED** [review]

File: eth/handler.go (hunk 3 above)
Comment (pratikspatil024) [reply eth/handler.go]: Oops, my bad. Updated, thanks!

**Pull request merged**